p2p: Restrict self-advertisements with privacy networks to avoid fingerprinting

mzumsande opened the pull request:
The current logic for self-advertisements works such that we detect as many local addresses as we can, and then, using the scoring matrix from `CNetAddr::GetReachabilityFrom()`, self-advertise with the address that fits best to our peer.
It is in general not hard for our peers to distinguish our self-advertisements from other addrs we send them, because we self-advertise every ~24h and because the first addr we send over a connection is likely our self-advertisement. 

`GetReachabilityFrom()` currently only takes into account actual reachability, but not whether we'd _want_ to announce our identity for one network to peers from other networks, which is not straightforward in connection with privacy networks. 

While the general approach is to prefer self-advertising with the address for the network our peer is on, there are several special situations in which we don't have one, and as a result could allow self-advertise other local addresses, for example:

A) We run i2p and clearnet, use `-i2pacceptincoming=0` (so we have no local i2p address), and we have a local ipv4 address. In this case, we'd advertise the ipv4 address to our outbound i2p peers.

B) Our `-discover` logic cannot detect any local clearnet addresses in our network environment, but we are actually reachable over clearnet. If we ran bitcoind clearnet-only, we'd always advertise the address our peer sees us with instead, and could get inbound peers this way. Now, if we also have an onion service running (but aren't using tor as a proxy for clearnet connections), we could advertise our onion address to clearnet peers, so that they would be able to connect our clearnet and onion identities. 

This PR tries to avoid these situations by 
1.) never advertising our local Tor or I2P address to peers from other networks.
2.) never advertising local addresses from non-anonymity networks to peers from Tor or I2P

Note that this affects only our own self-advertisements, the rules to forward other people's addrs are not changed.

[Edit] after Initial [discussion](https://github.com/bitcoin/bitcoin/pull/27411#issuecomment-1497176155): CJDNS is not being treated like Tor and I2P at least for now, because it has different privacy properties and for the practical reason that it has still very few bitcoin nodes.

dergoegge commented:
Concept ACK

Personally, I have given up a little bit on fighting fingerprinting (at least I've de-prioritized it) because at this point there are so many fingerprinting techniques that we should probably either give up or figure out a way to avoid them ~entirely by design.

> we could advertise our onion address to clearnet peers

But this is such an obvious fingerprint that we should address it.

vasild reviewed (commented):
Concept ACK

On the approach - this PR expands `CNetAddr::GetReachabilityFrom()` with some logic whether we want to advertise or not. Such a logic is already in `GetLocal()` and `CNetAddr::GetReachabilityFrom()` is simply about reachability. It would be good to keep this separation.

Is it not possible to achieve the same with something simpler, like the following?
```diff
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -164,12 +164,22 @@ bool GetLocal(CService& addr, const CNetAddr *paddrPeer)
     int nBestScore = -1;
     int nBestReachability = -1;
     {
         LOCK(g_maplocalhost_mutex);
         for (const auto& entry : mapLocalHost)
         {
+            // For privacy reasons, don't advertise our privacy-network address
+            // to other networks and don't advertise our other-network address
+            // to privacy networks.
+            const auto& local_addr = entry.first;
+            if ((local_addr.IsTor() && !paddrPeer->IsTor()) ||
+                (!local_addr.IsTor() && paddrPeer->IsTor()) ||
+                (local_addr.IsI2P() && !paddrPeer->IsI2P()) ||
+                (!local_addr.IsI2P() && paddrPeer->IsI2P())) {
+                continue;
+            }
             int nScore = entry.second.nScore;
```

vasild commented:
Or why not even expand this logic to all networks and delete `CNetAddr::GetReachabilityFrom()`? That is - advertise our X address only to peers from network X (for X being any of IPv4, IPv6, Tor, I2P, CJDNS).

mzumsande commented:
> Is it not possible to achieve the same with something simpler, like the following? (...)

We'd need one more `if` so that we don't stop advertising onion addresses to `127.0.0.1` peers (inbounds from tor).

But yes, I will try this alternative out.
Alternatively, I also thought of renaming `GetReachabilityFrom` to something like `GetAdvertisingScore` - since we don't use it anywhere else.

> Or why not even expand this logic to all networks and delete `CNetAddr::GetReachabilityFrom()`? That is - advertise our X address only to peers from network X (for X being any of IPv4, IPv6, Tor, I2P, CJDNS).

I think that would be too drastic: I don't see a reason to never advertise IPv4 addresses to IPv6 peers, and vice versa - for clearnet, better propagation seems more important than fingerprinting concerns. What do you think about CJDNS?

vasild commented:
Right, incoming Tor need special handling. That is already what `CNode::ConnectedThroughNetwork()` does. So it would look like:

```cpp
            // For privacy reasons, don't advertise our privacy-network address
            // to other networks and don't advertise our other-network address
            // to privacy networks.
            const Network our_net{entry.first.GetNetwork()};
            const Network peers_net{node.ConnectedThroughNetwork()};
            if (our_net != peers_net && 
                (our_net == NET_ONION || our_net == NET_I2P || 
                 peers_net == NET_ONION || peers_net == NET_I2P)) {
                continue;                                                  
            }                              
```
(and pass `CNode&` instead of `CNetAddr*`).

I am not sure about CJDNS. Normally I would say treat it like Tor and I2P (but I even had the heretic idea to do that with IPv4 and IPv6 too, so maybe I lean a bit to the paranoid side, don't listen to me too much). However there are very few CJDNS addresses right now on mainnet and if we only advertise our CJDNS address to CJDNS peers that could impede its adoption.

Ultimately, the perfect solution would be to broadcast our address exactly like we broadcast foreign addresses. So that a spying peer cannot distinguish which one is it. Right now we are more aggressive in broadcasting our address than broadcasting foreign addresses. To make it the same, either:
* Stop being more aggressive with our address, or
* Broadcast more aggressively also foreign addresses.

either one would be a serious change and would need some simulation to be justified. Just thinking aloud.

mzumsande committed:
net, refactor: pass CNode instead of CNetAddr to GetLocalAddress

Access to CNode will be needed in the following commits.

mzumsande committed:
net: restrict self-advertisements with privacy networks

Stop advertising
1) our i2p/onion address to peers from other networks
2) Local addresses of non-privacy networks to i2p/onion peers
Doing so could lead to fingerprinting ourselves.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

mzumsande committed:
net, refactor: pass reference for peer address in GetReachabilityFrom

The address of the peer always exists (because addr is a member of
CNode), so it was not possible to pass a nullptr before.
Also remove NET_UNKNOWN, which is unused now.

mzumsande commented:
> I am not sure about CJDNS. Normally I would say treat it like Tor and I2P (but I even had the heretic idea to do that with IPv4 and IPv6 too, so maybe I lean a bit to the paranoid side, don't listen to me too much). However there are very few CJDNS addresses right now on mainnet and if we only advertise our CJDNS address to CJDNS peers that could impede its adoption.

Makes sense to me - if bitcoin over CJDNS gets more popular we could revisit this later.
 
> Ultimately, the perfect solution would be to broadcast our address exactly like we broadcast foreign addresses. (...)

Yes, this would probably require an overhaul of the entire address relay mechanism, and I'm not sure if the benefits would be that great - after all, we don't really want to keep our address private, just not fingerprint ourselves in the most obvious way.

> Right, incoming Tor need special handling. That is already what `CNode::ConnectedThroughNetwork()` does. So it would look like: (...)

I think I will switch to this approach, keeping reachability and privacy concerns separate might be good. Will also try to add some unit test coverage for this area of code (there seems to be none so far), and then push an update in the next days.

mzumsande committed:
test: add unit test for local address advertising

mzumsande commented:
- Updated the PR to separate privacy checks from reachability checks.
- Passed a CNode to `GetLocal()` and replaced pointers with references in  `GetLocal()` and  `GetReachabilityFrom()` 
- Added a unit test

vasild commented on src/netaddress.cpp:
```diff
@@ -725,17 +725,15 @@ std::vector<unsigned char> CNetAddr::GetAddrBytes() const
 // and only used in GetReachabilityFrom
 static const int NET_UNKNOWN = NET_MAX + 0;
```
You may expand commit b3261389eadf050c0de590c2d9cd69ad9f62acff `net, refactor: pass reference for peer address in GetReachabilityFrom` to remove `NET_UNKNOWN` since it is unused after that commit.

mzumsande replied:
good catch, removed NET_UNKNOWN.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -905,4 +905,117 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+    BOOST_REQUIRE(addr_onion.IsTor());
+
+    CAddress addr_i2p;
+    BOOST_REQUIRE(addr_i2p.SetSpecial("UDHDrtrcetjm5sxzskjyr5ztpeszydbh4dpl3pl4utgqqw2v4jna.b32.I2P"));
```
nit: it is case-insensitive, so no point in using a mixture of lower case and upper case.

mzumsande replied:
fixed

vasild commented on src/test/net_tests.cpp:
```diff
@@ -905,4 +905,117 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+    BOOST_CHECK(GetLocalAddress(*peer_ipv6_tunnel) == addr_ipv4);
+    BOOST_CHECK(GetLocalAddress(*peer_teredo) == addr_ipv4);
+    BOOST_CHECK(GetLocalAddress(*peer_cjdns) == addr_ipv4);
+    BOOST_CHECK(GetLocalAddress(*peer_onion) == addr_empty);
```
nit: might as well ditch `addr_empty` and use

```suggestion
    BOOST_CHECK(!GetLocalAddress(*peer_onion).IsValid());
```

mzumsande replied:
done

vasild commented on src/test/net_tests.cpp:
```diff
@@ -905,4 +905,117 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+    RemoveLocal(addr_onion);
+    RemoveLocal(addr_i2p);
+
+    // local peers from all networks
```
```suggestion
    // local addresses from all networks
```

mzumsande replied:
fixed

vasild reviewed (approved):
ACK 6be8b6d8106210c6f44bc9c0124bfd1476d2952e

vasild reviewed (approved):
ACK 3f93ea39da26558cc731b6e4ecd8fa0152eaf94c

luke-jr commented:
>Note that this affects only our own self-advertisements, the rules to forward other people's addrs are not changed.

Would this make our own addr stand out too? Or could we still forward our own addr by coincidence outside of self-advertising?

vasild reviewed (approved):
ACK e7cf8657e1165ea5da3911a9e543837cd8938f97

mzumsande commented:
> > Note that this affects only our own self-advertisements, the rules to forward other people's addrs are not changed.
> 
> Would this make our own addr stand out too? Or could we still forward our own addr by coincidence outside of self-advertising?

Yes, we could still forward it: When we receive a gossipped address we choose 1-2 random peers (that stay the same for 24 hours, see [RelayAddress](https://github.com/bitcoin/bitcoin/blob/7952a5934a1e071ca24a51483d058174e7b6de43/src/net_processing.cpp#L2095)) and forward it to them if we don't think they already know it.
So if we receive our own address from someone else we could forward it to the chosen peer, even if that peer wasn't eligible for the original self-announcement because it's from another network.

luke-jr reviewed (approved):
utACK e7cf8657e1165ea5da3911a9e543837cd8938f97

Approach: Unsure about the refactoring, but ok

achow101 commented:
ACK e7cf8657e1165ea5da3911a9e543837cd8938f97

jonatack reviewed (commented):
Post-merge ACK

MarnixCroes commented:
Should this be added to the release notes?

mzumsande commented:
> Should this be added to the release notes?

Not sure. I think the impact on users isn't that strong (it fixes behavior in rather rare constellations), so I wouldn't have added it. But happy to add it, if others would like it included!
